File: src/main/java/at/searles/parsing/combinators/RecognizerThenParser.java
package at.searles.parsing.combinators;

import at.searles.parsing.*;
import at.searles.parsing.printing.ConcreteSyntaxTree;

public class RecognizerThenParser<T> implements Parser<T>, Recognizable.Then {
    private final Recognizer parent;
    private final Parser<T> parser;
    private final boolean allowParserBacktrack;

    public RecognizerThenParser(Recognizer parent, Parser<T> parser, boolean allowParserBacktrack) {
        this.parent = parent;
        this.parser = parser;
        this.allowParserBacktrack = allowParserBacktrack;
    }

    @Override
    public T parse(ParserStream stream) {
        long offset = stream.getOffset();

        long preStart = stream.getStart();
        long preEnd = stream.getEnd();

        if (!parent.recognize(stream)) {
            return null;
        }

        long start = stream.getStart();

        T t = parser.parse(stream);

        if (t == null) {
            if(stream.getOffset() != offset) {
                throwIfNoBacktrack(stream);
                stream.backtrackToOffset(offset);
            }

            stream.setStart(preStart);
            stream.setEnd(preEnd);
            return null;
        }

        stream.setStart(start);

        return t;
    }

    @Override
    public ConcreteSyntaxTree print(T t) {
        ConcreteSyntaxTree output = parser.print(t);

        // printTo in recognizer always succeeds.
        return output != null ? output.consLeft(parent.print()) : null;
    }

    @Override
    public String toString() {
        return createString();
    }

    @Override
    public Recognizable left() {
        return parent;
    }

    @Override
    public Recognizable right() {
        return parser;
    }

    @Override
    public boolean allowParserBacktrack() {
        return allowParserBacktrack;
    }
}

File: src/main/java/at/searles/parsing/combinators/RecognizerThenParser.kt
package at.searles.parsing.combinators

import at.searles.parsing.Parser
import at.searles.parsing.ParserStream
import at.searles.parsing.Recognizer
import at.searles.parsing.printing.ConcreteSyntaxTree

class RecognizerThenParser<T>(private val left: Recognizer, private val right: Parser<T>) : Parser<T> {

    override fun parse(stream: ParserStream): T? {
        if (!stream.recognize(left, true)) {
            return null
        }

        return stream.parse(right, false)
    }

    override fun recognize(stream: ParserStream): Boolean {
        return stream.recognize(left, true) && stream.recognize(right, false)
    }

    override fun print(item: T): ConcreteSyntaxTree? {
        val output = right.print(item)

        // printTo in recognizer always succeeds.
        return output?.consLeft(left.print())
    }

    override fun toString(): String {
        return "$left.then($right)"
    }

}